image_captioning/models/decoder.py

```diff
@@ -0,0 +1,120 @@
+        """
+        super().__init__(self)
+
+        self.hidden_units = hidden_units
+        self.attn_hidden_units = attn_hidden_units
+        self.vocab_size = vocab_size
+        self.embedding_dim = embedding_dim
+
+        self.attention = BahdanauAttention(attn_hidden_units)
+
+        # Convert word indices to embedding vectors.
+        self.embedding = tf.keras.layers.Embedding(input_dim=vocab_size,
+                                                   output_dim=embedding_dim)
+
+        self.lstm = tf.keras.layers.LSTM(units=self.hidden_units,
+                                         return_sequences=True,
+                                         return_state=True,
+                                         recurrent_initializer='glorot_uniform')
+
+        self.dense1 = tf.keras.layers.Dense(units=hidden_units)
+
+        # Final layer to convert into "scores" per each vocabulary.
+        self.dense2 = tf.keras.layers.Dense(units=vocab_size)
+
+    # pylint: disable=arguments-differ
+    def call(
+            self, prev_word: tf.Tensor, features: tf.Tensor,
+            prev_hidden_state: tf.Tensor
+    ) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor]:
+        """
+        Arguments:
+            prev_word: Sequence of indices of previous words,
+                used as ingredients to predict the next `word_seq_len` words.
+                `shape == (batch, word_seq_len)`.
+            features: Sequence of feature embeddings for the corresponding
+                image.
+                `shape == (batch, feature_seq_len, embedding_dim)`.
+            prev_hidden_state: Previous hidden state (`final_memory_state`).
+                `shape == (batch, hidden_units)`.
+
+        Returns:
+            seq_out: Output scores for each word vocabulary.
+                `shape == (batch, word_seq_len, vocab_size)`
+            final_memory_state: Final memory state for LSTM module.
+                `shape == (batch, hidden_units)`
+            attention_weights: Attention weights that were used to create
+                the context vector.
+                `shape == (batch, feature_seq_len, attn_weight=1)`
+        """
+        # context_vector.shape == (batch, embedding_dim)
+        # attention_weights.shape == (batch, feature_seq_len, attn_weight=1)
+        context_vector, attention_weights = self.attention(
+            features, prev_hidden_state)
+
+        # Create word embeddings
+        # x.shape == (batch, word_seq_len, embedding_dim)
+        word_embeddings = self.embedding(prev_word)
+
+        # Add a time axis to the context vector and repeat the context
+        # vector over the length of time so that it can be concatenated
+        # with `word_embeddings`.
+        # context_vector.shape == (batch, word_seq_len, embedding_dim)
+        context_vector = tf.broadcast_to(tf.expand_dims(context_vector, axis=1),
+                                         word_embeddings.shape)
```

Comment:
Don't know if we need this sequence to sequence handling case, but it's a nice touch.
May be useful if we wanna use this for VQA?